btx/processing/freqdir.py

```diff
@@ -0,0 +1,740 @@
+            n-x retrieved image segments of dimension end_index-start_index
+        """
+        self.imgsTracked.append((self.psi.counter, self.psi.counter + n))
+
+        bin_factor = self.bin_factor
+        downsample = self.downsample
+
+        # may have to rewrite eventually when number of images becomes large,
+        # i.e. streamed setting, either that or downsample aggressively
+        imgs = self.psi.get_images(n, assemble=False)
+
+        if downsample:
+            imgs = bin_data(imgs, bin_factor)
+
+        imgs = imgs[
+            [i for i in range(imgs.shape[0]) if not np.isnan(imgs[i : i + 1]).any()]
+        ]
+
+        num_valid_imgs, p, x, y = imgs.shape
+        formatted_imgs = np.reshape(imgs, (num_valid_imgs, p * x * y)).T
+
+        return formatted_imgs
+
+    def fetch_and_update_model(self, n):
+        """
+        Fetch images and update model.
+
+        Parameters
+        ----------
+        n : int
+            number of images to incorporate
+        """
+        img_batch = self.get_formatted_images(n)
+        self.john_update_model(img_batch)
+
+
+    def john_update_model(self, X):
+        """
+        Update matrix sketch with new batch of observations
+        
+        Parameters
+        ----------
+        X: ndarray
+            data to update matrix sketch with
+        """
+        _, numIncorp = X.shape
+        origNumIncorp = numIncorp
+        n = self.num_incorporated_images
+        q = self.ell
+        with TaskTimer(self.task_durations, "total update"):
+            if self.rank==0 and not self.merger:
+                print(
+                    "Factoring {m} sample{s} into {n} sample, {q} component model...".format(
+                        m=numIncorp, s="s" if numIncorp > 1 else "", n=n, q=q
+                    )
+                )
+            for row in X.T:
+                canRankAdapt = numIncorp > (self.ell + 15)
+                if self.nextZeroRow >= self.m:
+                    if self.increaseEll and canRankAdapt and self.rankAdapt:
+                        self.ell = self.ell + 10
+                        self.m = 2*self.ell
+                        self.sketch = np.vstack((*self.sketch, np.zeros((20, self.d))))
+                        self.increaseEll = False
+                        print("INCREASING RANK OF PROCESS {} TO {}".format(self.rank, self.ell))
+                    else:
+                        copyBatch = self.sketch[self.ell:,:].copy()
+                        self.john_rotate()
+                        if canRankAdapt and self.rankAdapt:
+                            reconError = np.sqrt(self.lowMemoryReconstructionErrorUnscaled(copyBatch))
+                            if (reconError > 0.08):
+                                self.increaseEll = True
+                self.sketch[self.nextZeroRow,:] = row 
+                self.nextZeroRow += 1
+                self.num_incorporated_images += 1
+                numIncorp -= 1
```

Comment:
I'd greatly appreciate some explanation about what is being done in the docstring: thank you!